### PR TITLE
Multistage build docker

### LINKED ISSUE #93: Reduce docker image size with multistage builds
Currently the image of `Nitro-Web` way too large (314MB). Many build dependencies are included which are irrelevant for production.

download 6

We would like to slim down the image size with multistage builds. We did that for the backend recenlty and the image size was reduced from >200MB to 110MB.
You can have a look into the Dockerfile of the backend to ~copy+paste~ get some inspiration:
https://github.com/Human-Connection/Nitro-Backend/blob/master/Dockerfile

### PR DESCRIPTION
**Pull request by mattwr18**
fixes #93

### CODE REVIEW

File: docker-compose.travis.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+version: "3.7"
+
+services:
+  backend:
+    image: humanconnection/nitro-web:builder
```
[the remark this answers is not in the document]
Comment (roschaefer) [reply docker-compose.travis.yml]: Do you know the reason for having yet another `docker-compose.yml` just for travis? Because of shared folders. Travis struggles to delete them after running the tests, thus silently failing our deployments. The reason why Travis cannot delete the files is that the `docker` user on Travis creates the files, not the user who is later trying to delete them.

File: .travis.yml [hunk 2]
```diff
@@ -18,8 +18,8 @@ before_install:
   - sudo mv docker-compose /usr/local/bin
 
 install:
-  - docker build --build-arg BUILD_COMMIT=$TRAVIS_COMMIT -t humanconnection/nitro-web .
-  - docker-compose -f docker-compose.yml up -d
+  - docker build --build-arg BUILD_COMMIT=$TRAVIS_COMMIT --target production -t humanconnection/nitro-web .
+  - docker-compose -f docker-compose.travis.yml up -d
```
[the remark this answers is not in the document]
Comment (mattwr18) [reply .travis.yml]: Great, I'll give this a go! I am exposing 3000 on line 34... although I believe it is waiting cause it is not spinning up properly, although, it is failing silently

[the remark this answers is not in the document]
Comment (roschaefer) [reply .travis.yml]: Are you sure? I believe with `docker-compose` you have to expose all the ports in the `.yml` file explicitly. (Although I never looked it up)

File: .travis.yml [hunk 3]
```diff
@@ -19,7 +19,7 @@ before_install:
 
 install:
   - docker build --build-arg BUILD_COMMIT=$TRAVIS_COMMIT --target production -t humanconnection/nitro-web .
-  - docker-compose -f docker-compose.yml -f docker-compose.travis.yml up -d
+  - docker-compose -f docker-compose.yml up -f docker-compose.travis.yml up -d
```
Comment (roschaefer) [inline .travis.yml]: no no, the `up` is there already :wink: @mattwr18

**Comment by roschaefer** [comment]
@mattwr18 great work, can you tell me how large the docker image is on your machine after this PR?

File: .travis.yml (hunk 3 above)
Comment (mattwr18) [reply .travis.yml]: ahh, right... I thought I had missed it yesterday from your review above... it's still failing the build though

**Comment by roschaefer** [comment]
@mattwr18 yes, you can add a repository and a tag. See our [build server](https://github.com/Human-Connection/Nitro-Web/blob/master/.travis.yml#L21) and the [deploy script](https://github.com/Human-Connection/Nitro-Web/blob/master/scripts/docker_push.sh#L3):
```
docker build [...] -t humanconnection/nitro-web:latest .
```
adds the repository and 
```
docker push humanconnection/nitro-web:latest
```
does the same for `docker push`.

**Comment by mattwr18** [comment]
Output from build, command `docker-compose -f docker-compose.yml -f docker-compose.travis.yml up -d`
```
...
Creating nitro-web_webapp_1
...
```
but unfortunately still getting this output
```
docker-compose exec webapp yarn run lint
No container found for webapp_1
The command "docker-compose exec webapp yarn run lint" exited with 1.
```

**Comment by mattwr18** [comment]
I've run the exact same commands as from the build locally and they have passed...

**Comment by mattwr18** [comment]
Size of the production image now
```
humanconnection/nitro-web       latest              aa9218a28cfd        5 minutes ago       319MB
```

**Comment by appinteractive** [comment]
Why not just copy all into the build image and trusting the dockerignore file? @mattwr18

**Comment by mattwr18** [comment]
that's a good idea @appinteractive, right now we were just concerned with the minimum to get it passing the build, but we can definitely explore that route. 

Currently, it's unexpectedly failing the build because of many linting errors, which we are not able to reproduce locally. 
https://travis-ci.com/Human-Connection/Nitro-Web/builds/96354499#L1006

File: Dockerfile [hunk 4]
```diff
@@ -27,6 +29,7 @@ RUN yarn run build
 
 FROM base as production
 ENV NODE_ENV=production
+RUN rm -rf ./nitro-web/components ./nitro-web/store
```
Comment (roschaefer) [inline Dockerfile]: @mattwr18 unfortunately, this adds another layer to the `production` stage and therefore it won't reduce the image size.

File: Dockerfile [hunk 5]
```diff
@@ -1,28 +1,38 @@
-FROM node:10-alpine
+FROM node:10-alpine as base
 LABEL Description="Web Frontend of the Social Network Human-Connection.org" Vendor="Human-Connection gGmbH" Version="0.0.1" Maintainer="Human-Connection gGmbH (developer@human-connection.org)"
 
 # Expose the app port
-EXPOSE 3000
-
+ARG BUILD_COMMIT
```
Comment (appinteractive) [inline Dockerfile]: You removed the port!

```suggestion
EXPOSE 3000
ARG BUILD_COMMIT
```

File: Dockerfile [hunk 6]
```diff
@@ -1,28 +1,38 @@
...
 
 # See: https://github.com/nodejs/docker-node/pull/367#issuecomment-430807898
 RUN apk --no-cache add git
 
-# Install Web Application
-COPY package.json .
-COPY yarn.lock .
-COPY styleguide/ ./styleguide
-RUN yarn install --production=false --frozen-lockfile --non-interactive
+COPY locales ./locales
```
Comment (appinteractive) [inline Dockerfile]: The copy part belongs into the builder where you copy stuff then over.

File: Dockerfile (hunk 4 above)
Comment (appinteractive) [reply Dockerfile]: Copy all into builder and then pick what you need out if that image by specifying the image (see backend build)

File: cypress/integration/common/steps.js [hunk 7]
```diff
@@ -1,6 +1,6 @@
 import { Given, When, Then } from 'cypress-cucumber-preprocessor/steps'
 import find from 'lodash/find'
-import { eq } from 'semver';
+import { eq } from 'semver'
```
Comment (appinteractive) [inline cypress/integration/common/steps.js]: Do we use that Import at all?

File: Dockerfile (hunk 4 above)
Comment (mattwr18) [reply Dockerfile]: > Why not just copy all into the build image and trusting the dockerignore file? @mattwr18

When you say copy all into the build image, what do you mean by that? if you look at this line https://github.com/Human-Connection/Nitro-Web/blob/9f8a292c7e39030a4fb05622c8830e5dbff967db/Dockerfile#L27 we are copying all; however, this all is only what has been copied from the host into the base image.

File: Dockerfile (hunk 5 above)
Comment (mattwr18) [reply Dockerfile]: we didn't remove the port, it is here https://github.com/Human-Connection/Nitro-Web/blob/9f8a292c7e39030a4fb05622c8830e5dbff967db/Dockerfile#L27... the `CMD` and `EXPOSE`  do not need to be in a particular line in the Dockerfile, they are not used during the build, they are only used 1) during deploy 2) when using `docker-compose up`

File: Dockerfile (hunk 6 above)
Comment (mattwr18) [reply Dockerfile]: Similar to my other comment above, we cannot copy anything in the `build-and-test` stage from host, it must first be copied from host to `base`, which we then use in both the `build-and-test` stage and `production` stage

File: cypress/integration/common/steps.js (hunk 7 above)
Comment (mattwr18) [reply cypress/integration/common/steps.js]: I'm not sure, as this is not code I wrote or am yet familiar with...

File: Dockerfile (hunk 4 above)
Comment (mattwr18) [reply Dockerfile]: > @mattwr18 unfortunately, this adds another layer to the `production` stage and therefore it won't reduce the image size.

I would have thought so, as it does add another layer, although, we checked the image size both before and after adding this line and did not observe a difference.

File: Dockerfile (hunk 6 above)
Comment (appinteractive) [reply Dockerfile]: Thats not true to my knowladge, you can copy files and folder from one image to another without the need to extend it. `COPY <src> <dest> --from=builder`

File: Dockerfile (hunk 5 above)
Comment (appinteractive) [reply Dockerfile]: Ahh allright I just saw the "removal" in the diff but missed that its now on the end of the file.

File: Dockerfile (hunk 6 above)
Comment (appinteractive) [reply Dockerfile]: For some reference: https://glebbahmutov.com/blog/making-small-docker-image/

**Comment by mattwr18** [comment]
> Yeah that’s because of failing config files for the Linter. Please try to copy the whole root dir and trust the dockerignore. After that you just copy the needed files to the production image.
> 
> That way the build will be more solid to change imho.

I understand that it is failing because it is failing the lint step; however, since it is not failing in the same way locally, the question is why is it failing in the build? Did you have a chance to look at the build? There are more than 30 errors in the linting, which only appear in the build, any idea why?

The issue with copying everything to base and then copying what we need from builder is that we are using base as our base image for production as well, so that would copy over everything to production and then additionally copy the directories we need from builder, which would not reduce the size of the image.

**Comment by mattwr18** [comment]
@appinteractive I have looked at the backend build, which is similar, but different in key ways. Some of the similarities, however, are that they both use `base` as the base image for `builder/build-and-test` and production stages, and they both do not copy everything from host to the base image, see https://github.com/Human-Connection/Nitro-Backend/blob/master/Dockerfile#L10

File: Dockerfile (hunk 4 above)
Comment (appinteractive) [reply Dockerfile]: Just as a note from my experience, you can’t copy stuff that’s ignored, even if you reference it one by one. Maybe that helps, was a catch for me.

File: Dockerfile (hunk 6 above)
Comment (mattwr18) [reply Dockerfile]: That is possibly a very nice solution; however, that is different from the Nitro-Backend, which you indicated should be used as a reference. Also, the link you shared does use the base image for both TEST and PROD... the only difference I guess is that they copy everything from host to the base image, which we can certainly try, but I would be surprised if reduced the size of the image drastically.

It's not really clear to me how these suggestions help to solve the failing build, but I can certainly give it a go, and if it create a larger image that reliably passes the build, it would be a good first step.

File: Dockerfile (hunk 5 above)
Comment (mattwr18) [reply Dockerfile]: yeah, no worries... I think that we could definitely do a refactor once we get it reliably passing the build... maybe I should put a [WIP] tag on the `PR`...

File: Dockerfile (hunk 4 above)
Comment (mattwr18) [reply Dockerfile]: That is my understanding as well; however, it's not clear to me that everything that should be in the dockerignore file actually is.

File: Dockerfile (hunk 6 above)
Comment (appinteractive) [reply Dockerfile]: I just try to give some input, might not be the best as I’m doing it from the go so I’m not involved 100%, sorry for that if it’s not helpful, just ignore it if it does not make sense.

**Comment by appinteractive** [comment]
> @appinteractive I have looked at the backend build, which is similar, but different in key ways. Some of the similarities, however, are that they both use `base` as the base image for `builder/build-and-test` and production stages, and they both do not copy everything from host to the base image, see https://github.com/Human-Connection/Nitro-Backend/blob/master/Dockerfile#L10

Alright than I mixed it up with some experiments I did for multistage builds some weeks ago.

File: Dockerfile (hunk 6 above)
Comment (mattwr18) [reply Dockerfile]: I think it very much is helpful! I am a bit stuck on why it's failing in the way it's failing, and having ideas of things to try is always welcome! thanks for taking the time to try to debug it... I will push it up and see what happens. Maybe we get lucky

**Comment by mattwr18** [comment]
So @appinteractive your suggestion to copy over everything in the base image has only made the image 20mb bigger and has seemed to help with the build(?) as there are less lint errors, only 2 now, but as a side effect the tests are failing :( 

I have been able to reproduce the errors locally by running those commands `docker-compose exec webapp yarn run (lint|test)` in `NODE_ENV=production`, I changed the `build-and-test` stage to use `NODE_ENV=test` and have run the command `docker-compose -f docker-compose.yml -f docker-compose.travis.yml up -d` locally and that resolved the errors, but it is still failing the build. It seems to still be being run in production mode... 

I have requested that Travis enable debug mode, which they have... however, I have tried to follow the instructions provided in the email from Travis, which I have successfully used in another AgileVentures project, but it is returning a "wrong credentials" response.

I have emailed travis to clarify whether the API token is specific to the Human Connection repos or it should be user specific. Maybe I don't have permissions to enter the build in debug mode.

File: cypress/integration/common/steps.js (hunk 7 above)
Comment (roschaefer) [reply cypress/integration/common/steps.js]: I removed it for now, we should see a failing cypress test if anything breaks

**Comment by mattwr18** [comment]
hey @appinteractive @roschaefer I finally got this passing the build by running the linting and testing in `NODE_ENV=test`, please let me know how you feel about this approach and if there is another, better get the build to pass.

File: .travis.yml [hunk 8]
```diff
@@ -18,17 +18,17 @@ before_install:
...
   - git -C "../Nitro-Backend" checkout $TRAVIS_BRANCH || echo "Branch \`$TRAVIS_BRANCH\` does not exist, falling back to \`master\`"
   - docker-compose -f ../Nitro-Backend/docker-compose.yml -f ../Nitro-Backend/docker-compose.travis.yml up -d
   - yarn global add cypress wait-on
   - yarn add cypress-cucumber-preprocessor
 
 script:
-  - docker-compose exec webapp yarn run lint
-  - docker-compose exec webapp yarn run test
+  - docker-compose exec -e NODE_ENV=test webapp yarn run lint
+  - docker-compose exec -e NODE_ENV=test webapp yarn run test
```
Comment (roschaefer) [inline .travis.yml]: @mattwr18 I would put `NODE_ENV=test` in the Dockerfile to the `build-and-test` stage.

File: Dockerfile [hunk 9]
```diff
@@ -1,28 +1,32 @@
...
     && cd .. \
     && yarn run styleguide:build
-
 COPY . .
-RUN ["yarn", "run", "build"]
+RUN yarn run build
+
+FROM base as production
+ENV NODE_ENV=production
+COPY --from=build-and-test ./nitro-web/node_modules ./node_modules
```
Comment (roschaefer) [inline Dockerfile]: @mattwr18 can you try to compress the docker image and `yarn install` the dependencies instead of copying them from `build-and-test`?

File: .travis.yml (hunk 8 above)
Comment (mattwr18) [reply .travis.yml]: Tried that first, it doesn't pass the build that way... see commit `890d224`

File: Dockerfile (hunk 9 above)
Comment (mattwr18) [reply Dockerfile]: what do you mean compress the docker image? you had suggested the approach of running `yarn install` in the production image to avoid copying over the `devDependencies` earlier, but it doesn't pass the build because of this dependency `nuxt-sass-resources-loader`, maybe it is incorrectly in `devDependencies`?

Comment (appinteractive) [reply Dockerfile]: From their [github page](https://github.com/anteriovieira/nuxt-sass-resources-loader): `"Note that installing as a dev dependency --save-dev or -D will not work correctly."` 
So yes It can`t be only in the dev devDepemdencies!

File: Dockerfile [hunk 10]
```diff
@@ -1,28 +1,32 @@
...
 
 # See: https://github.com/nodejs/docker-node/pull/367#issuecomment-430807898
 RUN apk --no-cache add git
 
-# Install Web Application
-COPY package.json .
-COPY yarn.lock .
-COPY styleguide/ ./styleguide
-RUN yarn install --production=false --frozen-lockfile --non-interactive
+COPY . .
```
Comment (appinteractive) [inline Dockerfile]: To save more space, we should not copy everything into the base but into the build-and-test image, and then copying only the stuff needed. Thats also because the styleguide is quite heavy and I think only the dist directory is needed. 

But that could by done also in another pr.

File: Dockerfile [hunk 11]
```diff
@@ -1,28 +1,32 @@
...
     && yarn run styleguide:build
-
 COPY . .
-RUN ["yarn", "run", "build"]
+RUN yarn run build
+
+FROM base as production
+ENV NODE_ENV=production
+COPY --from=build-and-test ./nitro-web/node_modules ./node_modules
+COPY --from=build-and-test ./nitro-web/plugins ./plugins
```
Comment (appinteractive) [inline Dockerfile]: Why do you copy the plugins and static files over when you already copied everything in the base image?

File: Dockerfile (hunk 10 above)
Comment (mattwr18) [reply Dockerfile]: From what I observed entering the `build-and-test` image, we cannot copy over anything that was not first copied over by the base image.

File: Dockerfile (hunk 11 above)
Comment (mattwr18) [reply Dockerfile]: Good point, carry over from before your above suggestion.

**Comment by mattwr18** [comment]
@appinteractive @roschaefer I have gotten this build to pass reliably again and cleaned it up even more in the process, taking into account your suggestions. It is not an ideal situation, however. 

The reason it was failing the build before @roschaefer is because for some reason the commands `docker-compose exec webapp yarn run lint` and `yarn run test` are being run from the production image, which after having removed the copy node_modules from build and test, does not include the devDependencies jest or eslint necessary to run the commands. 

This is also the reason I needed to add the `-e NODE_ENV=test`... I was having a look at how the `docker-compose -f docker-compse.yml -f docker-compose.travis.yml up -d` command works and it seems like the Docker docs encourage to have a separate yml for production. See https://docs.docker.com/compose/extends/#different-environments and https://docs.docker.com/compose/production/ @roschaefer

**Comment by roschaefer** [comment]
For the record:
screenshot 21

**Comment by roschaefer** [comment]
@mattwr18 I'm done, I believe this is a bug with `docker-compose.yml`. My latest reverted change is an equivalent refactoring. I can't believe why the build was failing. Nevertheless, I think this PR is an improvement. I'll merge

**Comment by roschaefer** [comment]
@mattwr18 :tada: 
screenshot 22

**Comment by mattwr18** [comment]
Wow!! what happened!? I thought it was around 315MB, still an big improvement, but nothing the 114MB!!! 
This is great!! I want to get more experience with this by multi-staging all the Dockerfiles I helped create!! Thanks for your help @appinteractive @roschaefer @ksinkar :)

**Comment by roschaefer** [comment]
@mattwr18 I assume it comes from the `node_modules` and the built files in `styleguide` that are not included. I still believe it can be compressed even more by not including the development dependencies. But after spending hours of try/error and being confused by the behaviour of docker-compose I leave that for some point in the future.
